### PR TITLE
Fix some issues with Docker Compose environment

### PR DESCRIPTION
**Pull request by barrywhart**
Fixes some issues noted by a contributor while working on PR #4198

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: plugins/sqlfluff-templater-dbt/test/conftest.py [hunk 1]
```diff
@@ -10,4 +11,4 @@ def dbt_flags():
     # some test code in core/dbt/parser/models.py, ModelParser. render_update().
     # We've seen occasional runtime errors from that code:
     # TypeError: cannot pickle '_thread.RLock' object
-    dbt.flags.USE_EXPERIMENTAL_PARSER = True
+    os.environ["DBT_USE_EXPERIMENTAL_PARSER"] = "True"
```
Comment (barrywhart) [inline plugins/sqlfluff-templater-dbt/test/conftest.py:14]: Second attempt at the `TypeError: cannot pickle '_thread.RLock' object` fix I mentioned in Slack. I saw the error again in the build. Apparently the original code wasn't the right way to set the flag. I "tested" this change by temporarily adding some assertions in the templater, dbt, and osmosis code. The flag appeared to be set correctly in all cases. 🤞 that this fixes it "for real".

File: plugins/sqlfluff-templater-dbt/docker/init [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+pip install --no-dependencies -e . -e plugins/sqlfluff-templater-dbt
```
Comment (barrywhart) [inline plugins/sqlfluff-templater-dbt/docker/init:2]: Need to do all this stuff once the container is running in order for the tests to work in Docker.

File: plugins/sqlfluff-templater-dbt/docker/shell [hunk 3]
```diff
@@ -1,3 +1,3 @@
 #!/usr/bin/env bash
 my_path="$( cd "$(dirname "$0")"; pwd -P)"
-docker-compose -f ${my_path}/docker-compose.yml exec app bash
+docker compose -f ${my_path}/docker-compose.yml exec app bash
```
Comment (barrywhart) [inline plugins/sqlfluff-templater-dbt/docker/shell:3]: The `docker-compose` command is deprecated and IIUC has been removed from some recent versions of Docker.

**Review by WillAyd — APPROVED** [review]
lgtm

File: tox.ini [hunk 4]
```diff
@@ -14,7 +14,7 @@ setenv =
     dbt{100,130}: PIP_CONSTRAINT=constraints/{envname}.txt
 allowlist_externals =
     make
-pip_pre = true
+pip_pre = false
```
Comment (barrywhart) [inline tox.ini:17]: Unrelated change discussed in maintainers' Slack channel: Don't install pre-release packages. It destabilizes the build; there's no good reason for doing this AFAIK.

**Review by alanmcruickshank — APPROVED** [review]
I don't get all of what we're doing in the docker init and docker shell files, but the rest seems great. 👍
